From patches Mon Sep 17 00:00:00 2001
From: Cornelia Huck <cornelia.huck@de.ibm.com>
To: Alexander Graf <agraf@suse.de>
Cc: Christian Borntraeger <borntraeger@de.ibm.com>,
 Cornelia Huck <cornelia.huck@de.ibm.com>, qemu-devel@nongnu.org,
 Dominik Dingel <dingel@linux.vnet.ibm.com>
Date: Tue, 23 Apr 2013 14:37:14 +0200
Message-id: 1366720634-13184-5-git-send-email-cornelia.huck@de.ibm.com
Subject: [PATCH 4/4] s390-ccw.img: Get queue config from host.

Ask the host about the configuration instead of guessing it.

Signed-off-by: Cornelia Huck <cornelia.huck@de.ibm.com>
---
 pc-bios/s390-ccw/virtio.c | 10 +++++++---
 pc-bios/s390-ccw/virtio.h |  5 +++++
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/pc-bios/s390-ccw/virtio.c b/pc-bios/s390-ccw/virtio.c
index 8e37dce..6b28aab 100644
--- a/pc-bios/s390-ccw/virtio.c
+++ b/pc-bios/s390-ccw/virtio.c
@@ -261,17 +261,21 @@ int virtio_read(ulong sector, void *load_addr)
 void virtio_setup_block(struct subchannel_id schid)
 {
     struct vq_info_block info;
+    struct vq_config_block config = {};
 
     virtio_reset(schid);
 
-    /* XXX need to fetch the 128 from host */
-    vring_init(&block, 128, (void*)(100 * 1024 * 1024),
+    config.index = 0;
+    if (run_ccw(schid, CCW_CMD_READ_VQ_CONF, &config, sizeof(config))) {
+        virtio_panic("Could not get block device configuration\n");
+    }
+    vring_init(&block, config.num, (void*)(100 * 1024 * 1024),
                KVM_S390_VIRTIO_RING_ALIGN);
 
     info.queue = (100ULL * 1024ULL* 1024ULL);
     info.align = KVM_S390_VIRTIO_RING_ALIGN;
     info.index = 0;
-    info.num = 128;
+    info.num = config.num;
     block.schid = schid;
 
     if (!run_ccw(schid, CCW_CMD_SET_VQ, &info, sizeof(info))) {
diff --git a/pc-bios/s390-ccw/virtio.h b/pc-bios/s390-ccw/virtio.h
index a33199d..86fdd57 100644
--- a/pc-bios/s390-ccw/virtio.h
+++ b/pc-bios/s390-ccw/virtio.h
@@ -53,6 +53,11 @@ struct vq_info_block {
     u16 num;
 } __attribute__((packed));
 
+struct vq_config_block {
+    u16 index;
+    u16 num;
+} __attribute__((packed));
+
 struct virtio_dev {
     struct virtio_dev_header *header;
     struct virtio_vqconfig *vqconfig;
-- 
1.7.12.4


